### PR TITLE
HCC Runtime optimizations

### PR DESCRIPTION
**Pull request by bensander**
Collection of HCC runtime optimizations and improvements - see commit history for details.

Passes HIP dtests and tested in larger MI framework.

Not sure on failure to automatically merge?

### CODE REVIEW

**Comment by whchung** [comment]
@bensander your branch doesn't seem to track the tip of HCC. could you rebase and submit once again?

**Comment by bensander** [comment]
Done!

**Comment by bensander** [comment]
@whchung can you review?

File: lib/hsa/mcwamp_hsa.cpp [hunk 1]
```diff
@@ -1230,12 +1240,21 @@ class HSAQueue final : public KalmarQueue
...
+
+        bool foundFirstValidOp = false;
+
         for (int i = asyncOps.size()-1; i >= 0;  i--) {
             if (asyncOps[i] != nullptr) {
                 auto asyncOp = asyncOps[i];
+                if (!foundFirstValidOp) {
+                    hsa_signal_t sig =  *(static_cast <hsa_signal_t*> (asyncOp->getNativeHandle()));
+                    assert(sig->handle != 0);
+                    foundFirstValidOp = false;
```
Comment (scchan) [inline lib/hsa/mcwamp_hsa.cpp]: should set to "true" here?

**Review by scchan — CHANGES_REQUESTED** [review]

File: lib/hsa/mcwamp_hsa.cpp [hunk 2]
```diff
@@ -1084,21 +1088,25 @@ class HSAQueue final : public KalmarQueue
...
                     << std::endl);
 
 
 
-        if (asyncOps.size() >= MAX_INFLIGHT_COMMANDS_PER_QUEUE) {
+        if (!drainingQueue_ && (asyncOps.size() >= MAX_INFLIGHT_COMMANDS_PER_QUEUE-1)) {
             DBOUT(DB_WAIT, "*** Hit max inflight ops asyncOps.size=" << asyncOps.size() << ". op#" << op->getSeqNum() << " force sync\n");
             DBOUT(DB_RESOURCE, "*** Hit max inflight ops asyncOps.size=" << asyncOps.size() << ". op#" << op->getSeqNum() << " force sync\n");
 
+            drainingQueue_ = true;
```
Comment (whchung) [inline lib/hsa/mcwamp_hsa.cpp:1100]: do we need a mutex here?

File: lib/hsa/mcwamp_hsa.cpp (hunk 1 above)
Comment (whchung) [reply lib/hsa/mcwamp_hsa.cpp]: this line indeed seems to be incorrect.

**Review by whchung — CHANGES_REQUESTED** [review]

File: lib/hsa/mcwamp_hsa.cpp (hunk 1 above)
Comment (bensander) [reply lib/hsa/mcwamp_hsa.cpp]: Yep.  Pushed a fix.

File: lib/hsa/mcwamp_hsa.cpp (hunk 2 above)
Comment (bensander) [reply lib/hsa/mcwamp_hsa.cpp:1100]: I believe HCC queues are not thread-safe by design, so no mutex is required here.

**Review by whchung — APPROVED** [review]

**Pull request merged**